firedrake/assign.py

```diff
@@ -131,11 +131,11 @@ def __init__(self, assignee, expression, subset=None):
             expression = expression.function
         expression = as_ufl(expression)
 
-        if not all(c.function_space() == assignee.function_space()
+        if not all(c.ufl_element() == assignee.ufl_element()
```

Comment:
I still have concerns about users trying to add functions on different meshes, but agree this shouldn't prevent this fix going in. We should open an issue with an MFE.

Reply:
I've pushed a commit that does this.